File: src/1-500/004/MedianofTwoSortedArrays.py
from typing import List


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        index1 = 0
        index2 = 0
        mergedArr = []
        while index1 < len(nums1) and index2 < len(nums2):
            if nums1[index1] > nums2[index2]:
                mergedArr.append(nums2[index2])
                index2 += 1
            elif nums1[index1] < nums2[index2]:
                mergedArr.append(nums1[index1])
                index1 += 1
            else:  # equal
                mergedArr.append(nums1[index1])
                mergedArr.append(nums2[index2])
                index1 += 1
                index2 += 1

        if index1 != len(nums1):
            while index1 < len(nums1):
                mergedArr.append(nums1[index1])
                index1 += 1

        if index2 != len(nums2):
            while index2 < len(nums2):
                mergedArr.append(nums2[index2])
                index2 += 1

        mergedLen = len(mergedArr)
        if mergedLen % 2 == 0:  # even length
            return float(mergedArr[mergedLen // 2] + mergedArr[mergedLen // 2 - 1]) / 2
        else:
            return float(mergedArr[(mergedLen - 1) // 2])

solution = Solution()
print(solution.findMedianSortedArrays([1, 3], [2]))

File: src/1-500/004/MedianOfTwoSortedArrays.java
class Solution:
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
		int x = nums1.length;
		int y = nums2.length;
		
		if (x > y) {
			return findMedianSortedArrays(nums2, nums1);
		}
		
		int low = 0;
		int high = x;
		
		while (low <= high) {
			int partitionX = (low + high) / 2;
			int partitionY = (x+y+1)/2 - partitionX;
			
			int maxLeftX =  partitionX == 0 ? Integer.MIN_VALUE : nums1[partitionX-1]; 
			int minRightX =  partitionX == x ? Integer.MAX_VALUE : nums1[partitionX]; 
			
			int maxLeftY =  partitionY == 0 ? Integer.MIN_VALUE : nums2[partitionY-1]; 
			int minRightY =  partitionY == y ? Integer.MAX_VALUE : nums2[partitionY];
			
			if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
				if ((x+y)%2 == 0) {
					return (Math.max(maxLeftX,  maxLeftY) + Math.min(minRightX,  minRightY)) / 2.0;
				}
				else {
					return Math.max(maxLeftX,  maxLeftY);
				}
			}
			else if (maxLeftX <= minRightY) {
				low = partitionX + 1;
			}
			else {
				high = partitionX - 1;
			}
		}
		
		// never reach here
		return -1.0;
    }
}